### ISSUE TITLE
unconditional_recursion lint false positive with specialization

### ISSUE THREAD

**Issue by comex** [rust-lang/rust#56866]
In this example, a `default fn` in a trait impl unconditionally calls the same fn on a different type.  That type is covered by the same impl containing the `default fn`, but it's also covered by a specialized impl that overrides that fn, hence there is not actually infinite recursion.

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=69a389b0d71c9dee3fa55398c481f93)

```rust
#![feature(specialization)]

trait Foo {
    fn foo();
}

impl<T> Foo for T {
    default fn foo() {
        <Box<T> as Foo>::foo();
    }
}

impl Foo for Box<i32> {
    fn foo() {}
}

fn main() {
    i32::foo();
}
```
Output:
```
warning: function cannot return without recursing
 --> src/main.rs:8:5
  |
8 |     default fn foo() {
  |     ^^^^^^^^^^^^^^^^ cannot return without recursing
9 |         <Box<T> as Foo>::foo();
  |         ---------------------- recursive call site
  |
  = note: #[warn(unconditional_recursion)] on by default
  = help: a `loop` may express intention better if this is on purpose
```

**Comment by bluss** [comment]
If you switch the specialization to be this:

```rust

impl<T> Foo for Box<T> {
    fn foo() {}
}
```

then the lint disappears. So isn't the compiler right all the way? It flags your implementation and for all T!=i32 it will unconditionally recurse.
